src/buildstream_plugins/sources/cargo.py

```diff
@@ -148,7 +148,25 @@ def stage(self, directory):
         try:
             mirror_file = self._get_mirror_file()
             with tarfile.open(mirror_file) as tar:
-                tar.extractall(path=directory)
+
+                def is_within_directory(directory, target):
+                    abs_directory = os.path.abspath(directory)
+                    abs_target = os.path.abspath(target)
+
+                    prefix = os.path.commonprefix([abs_directory, abs_target])
+
+                    return prefix == abs_directory
+
+                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
+
+                    for member in tar.getmembers():
+                        member_path = os.path.join(path, member.name)
+                        if not is_within_directory(path, member_path):
```

Comment:
I believe we could simplify the logic with `pathlib` here:

```suggestion
                def safe_extract(tar, path=".", members=None, *, numeric_owner=False):
                    path = Path(path).resolve()
                    
                    for member in tar.getmembers():
                        member_path = path.joinpath(member.name).resolve()
                        if path not in member_path.parents:
```